### PR TITLE
ivy.el: ivy-done: Pass input arg to REQUIRE-MATCH function

### PR DESCRIPTION
**Pull request by ianyepan**
The completing-read spec states that when REQUIRE-MATCH is a function, it should be called with the input as the argument:

"a function, which will be called with the input as the argument. If the function returns a non-nil value, the minibuffer is exited with that argument as the value."

Previously, ivy-done called `(funcall require-match)` with zero arguments, causing wrong-number-of-arguments errors when packages (e.g., Magit during branch creation) correctly pass a 1-argument validation function as REQUIRE-MATCH.

Changes:

- Add a new cond clause that calls `(funcall require-match ivy-text)` and exits via ivy--done when the function returns non-nil (valid)
- Remove the old clause that incorrectly called the function with no args and always prompted "(confirm)" regardless of validity
- Invalid input (function returns nil) naturally falls through to the `t` clause which shows "(match required)"

Context:

- Magit recently started passing a lambda as REQUIRE-MATCH to validate branch names in magit-branch--read-name
- This is correct per the completing-read contract but exposed the bug in Ivy's ivy-done

## Testing

Tested on Redhat-based Linux, macOS, and WSL2 Ubuntu. Confirms that the patch resolved the regression (or rather, bug exposure) by Magit's update.

### CODE REVIEW

**Comment by ianyepan** [comment]
This PR will fix issue https://github.com/abo-abo/swiper/issues/3082

**Comment by ianyepan** [comment]
cc @basil-conto if you're free to have a look. Until this PR is merged, any users that use ivy and magit together would experience breakage when executing commands like magit's branch create.

**Comment by basil-conto** [comment]
Thanks for the ping, but I'm sorry: I cannot look at this at the moment, while I wait for my new employer to disclaim rights to any changes I make to Emacs under copyright assignment to the Free Software Foundation.

In fact I'm ashamed to say that I have been sitting on a fix for this exact issue for over a year now, but didn't get around to finishing it up :/.

**Comment by ianyepan** [comment]
Hi @basil-conto, appreciate the quick reply, and I hope you and your new employer soon sort things out the way you wish! In the meantime, if there's any other collaborator with merge access that we could ping feel free to drop their names here for a quick review.

Otherwise let's wait until your employer disclaim rights so you could review the PR.

Best of luck,
Ian

**Comment by jrimmer-bendhealth** [comment]
I've taken to just disabling ivy-mode before I create branches in magit, then re-enabling afterwards, as a workaround.

It's lame, but I don't create branches more often than once or twice a day, so... I survive. :)

**Comment by basil-conto** [comment]
In the meantime, has the fix for magit/magit#5581 worked around this issue?

**Comment by basil-conto** [comment]
Thanks for your patience. Everything is sorted on my end now, so I aim to have a look at this and install it later this week.

**Comment by ianyepan** [comment]
Thanks @basil-conto - glad to hear that!

**Comment by ianyepan** [comment]
Hi @basil-conto , gentle ping. Let me know what you think about the PR. Thx!

**Comment by basil-conto** [comment]
Thanks for the ping. I started by trying to add some regression tests for this and naturally fell down several rabbit holes [fixing issues in the current test harness](https://github.com/basil-conto/swiper/commit/ab195babfa13df9c7a05407ac2c3b9019c6ef373) as well as Ivy's core completion support (and I overestimated how much free time I'd have).

I intend to park the increasing number of deficiencies I'm finding and just install this patch before continuing on my never-ending journey, later today or tomorrow.

_Update:_ from 8th August onwards, sorry :/

File: ivy.el [hunk 1]
```diff
@@ -749,11 +749,12 @@ candidate, not the prompt."
...
             ((string= " (confirm)" ivy--prompt-extra)
              (ivy--done ivy-text))
-            ((or (and (memq (ivy-state-collection ivy-last)
-                            '(read-file-name-internal internal-complete-buffer))
-                      (eq confirm-nonexistent-file-or-buffer t))
-                 (and (functionp require-match)
-                      (setq require-match (funcall require-match))))
+            ((and (functionp require-match)
+                  (funcall require-match ivy-text))
+             (ivy--done ivy-text))
```
Comment (basil-conto) [inline ivy.el:754]: `ivy-done` seems to use two different sources of 'user input': `ivy-text` and `ivy--input`. I don't know why they both exist, nor why or when they'd differ. My understanding goes only as far as `ivy-text` being a variable that has to be updated by Ivy as part of the `post-command-hook` command loop, whereas `ivy--input` reflects the current minibuffer contents independently of the command loop.

Until I understand them better, I suggest you pass `input` in place of `ivy-text` as the argument to `require-match`, and keep `ivy-text` as the argument to `ivy--done`, for consistency with how these variables are used in the rest of `ivy-done`.

Thoughts?

**Review by basil-conto — COMMENTED** [review]
Thank you very much for both the patch(es) and your patience with me!

> * Invalid input (function returns nil) naturally falls through to the `t` clause which shows "(match required)"

Is this strictly true?  It seems possible to me that the `require-match` predicate returns `nil`, and then the subsequent `confirm-nonexistent-file-or-buffer` condition succeeds, in which case the "match required" form isn't reached.

**Comment by basil-conto** [comment]
Just a heads up, I am close to a complete solution for this and other require-match edge cases. The main thing remaining to resolve is the selectable prompt logic in conjunction with require-match, which I hope to complete over this weekend. Other issues like #3063 may require a deeper redesign so I won't let that block this fix.

**Comment by basil-conto** [comment]
> Just a heads up, I am close to a complete solution for this and other require-match edge cases. The main thing remaining to resolve is the selectable prompt logic in conjunction with require-match, which I hope to complete over this weekend. Other issues like #3063 may require a deeper redesign so I won't let that block this fix.

Turns out the selectable prompt logic was pretty straightforward: just needed to remove the restriction on selecting the prompt when `require-match` is non-`nil`, and add exhaustive regression tests.

What's turned out to be more complicated and is holding me up is file name completion and its many edge cases, both in vanilla completion and with Ivy. Again I'm trudging through old issues, adding regression tests, reevaluating past decisions in Ivy with an eye towards consistency with vanilla completion, and fixing the implementation w.r.t. `require-match` in particular. Can't say it's fun. ;(
